dpdk/src/ring/mod.rs

```diff
@@ -0,0 +1,160 @@
+        /// TODO: figure out why musl builds don't expose E_RTE_NO_CONFIG
+        /// likely a config error for bindgen
+        // use dpdk_sys::_bindgen_ty_4::E_RTE_NO_CONFIG;
+        const E_RTE_NO_CONFIG: u64 = 1002;
+        use err::RingCreateErr::*;
+        let params = params.validate().map_err(InvalidArgument)?;
+        let name = CString::new(params.name())
+            .unwrap_or_else(|_| unreachable!("null characters already excluded"));
+        let socket_id = socket::SocketId::try_from(params.0.socket_preference).map_err(|e| {
+            UnableToDetermineNumaNode {
+                params: params.0.clone(),
+                code: e,
+            }
+        })?;
+
+        /// TODO: expose ring SP vs MC flags from dpdk-sys
+        // For now, 0x1 | 0x2 yields a single-producer, single-consumer queue
+        const FLAGS: c_uint = 0x1 | 0x2;
+        let inner = match NonNull::new(unsafe {
+            dpdk_sys::rte_ring_create(
+                name.as_ptr(),
+                params.size() as c_uint,
+                socket_id.0 as c_int,
+                FLAGS,
+            )
+        }) {
+            None => {
+                let errno = Errno::from(unsafe { dpdk_sys::rte_errno_get() });
+                if errno.0 == E_RTE_NO_CONFIG as i32 {
+                    return Err(NoConfig(params.0));
+                }
+                return match ErrorCode::parse_errno(errno) {
+                    ErrorCode::Standard(StandardErrno::InvalidArgument) => Err(InvalidArgument(
+                        err::InvalidArgument::SizeNotPowerOfTwo(params.0),
+                    )),
+                    ErrorCode::Standard(StandardErrno::NoSpaceLeftOnDevice) => {
+                        Err(NotEnoughMemZones(params.0))
+                    }
+                    ErrorCode::Standard(StandardErrno::FileExists) => Err(MemZoneExists(params.0)),
+                    ErrorCode::Standard(StandardErrno::NoMemory) => {
+                        Err(UnableToAllocateMemZone(params.0))
+                    }
+                    code => Err(UnexpectedErrno {
+                        code,
+                        params: params.0,
+                    }),
+                };
+            }
+            Some(ring_ptr) => ring_ptr,
+        };
+        Ok(Self {
+            inner,
+            params,
+            marker: PhantomData,
+            marker2: PhantomData,
+        })
+    }
+}
+
+pub mod err {
+    use crate::ring::Params;
+    use errno::ErrorCode;
+
+    #[derive(thiserror::Error, Debug)]
+    pub enum InvalidArgument {
+        #[error("size must be a power of two ({size} given)", size=.0.size)]
+        SizeNotPowerOfTwo(Params),
+        #[error("ring name must be ASCII")]
+        NameNotAscii(Params),
+        #[error("Only alphanumeric ring names are supported (may contain -, _, and .)")]
+        IllegalCharacters(Params),
+        #[error("Name too long (max is 127 ASCII characters)")]
```

Comment:
Nit: Lowercase/uppercase consistency (I personally prefer using an upper case here, but mostly it would be nice to remain consistent between the different error messages).

Same below in `RingCreateErr`.

Reply:
Good point.  The rust convention is lower case but my hands tend to automatically capitalize the first letter when I type a sentence.  Will fix

Reply:
resolved